### PR TITLE
Unify reg nfs/v1

### PR DESCRIPTION
**Pull request by inashivb**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/3315

### CODE REVIEW

File: rust/src/nfs/nfs.rs [hunk 1]
```diff
@@ -1799,57 +1855,93 @@ pub extern "C" fn rs_nfs_probe_ms(
...
+    }
+
     let slice: &[u8] = build_slice!(input, len as usize);
     SCLogDebug!("rs_nfs_probe: running probe");
-    return nfs_probe(slice, direction);
+    match nfs_probe(slice, direction) {
+        1 => { return unsafe { ALPROTO_NFS } },
+        -1 => { return unsafe { ALPROTO_FAILED } },
+        _ => { return ALPROTO_UNKNOWN },
+    }
```
Comment (jasonish) [inline rust/src/nfs/nfs.rs:1897]: In a block like this you should be able to remove the `return` statements, this will keep code analyzers happy and is more idiomatic Rust.

**Comment by jasonish** [comment]
I noticed the second commit does quite a bit more than just change the function signatures as mentioned in the commit message.  So I think some more explanation is required there.  Or if these extra additions are actually fixing issues maybe separate commits.

For example, `rs_nfs_parse_request` is now setting file flags, but isn't in master.

**Comment by inashivb** [comment]
> I noticed the second commit does quite a bit more than just change the function signatures as mentioned in the commit message. So I think some more explanation is required there. Or if these extra additions are actually fixing issues maybe separate commits.
> 
> For example, `rs_nfs_parse_request` is now setting file flags, but isn't in master.

Yes, it is to make it consistent with what C fn does so the C fn can be removed easily in the next commit but it indeed requires a separate commit of its own w some explanation.

File: rust/src/nfs/nfs.rs (hunk 1 above)
Comment (inashivb) [reply rust/src/nfs/nfs.rs:1897]: OK. Thank you!

File: rust/src/nfs/nfs.rs [hunk 2]
```diff
@@ -39,6 +40,9 @@ use crate::nfs::nfs3_records::*;
 
 pub static mut SURICATA_NFS_FILE_CONFIG: Option<&'static SuricataFileContext> = None;
 
+pub const NFS_MIN_FRAME_LEN: u16 = 32;
+
+static mut ALPROTO_NFS: AppProto = ALPROTO_UNKNOWN;
```
Comment (inashivb) [inline rust/src/nfs/nfs.rs:45]: Move consts to separate commit

File: rust/src/nfs/nfs.rs [hunk 3]
```diff
@@ -1359,7 +1363,7 @@ impl NFSState {
     }
     fn setfileflags(&mut self, direction: u8, flags: u16) {
         SCLogDebug!("direction: {}, flags: {}", direction, flags);
-        if direction == 1 {
+        if direction == 1 {        // TODO change toclient
```
Comment (inashivb) [inline rust/src/nfs/nfs.rs:1366]: Resolve TODO

File: rust/src/nfs/nfs.rs [hunk 4]
```diff
@@ -1387,14 +1391,23 @@ pub extern "C" fn rs_nfs_state_free(state: *mut std::os::raw::c_void) {
...
+                                       ) -> AppLayerResult
 {
+    let state = cast_pointer!(state, NFSState);
+    let flow = cast_pointer!(flow, Flow);
+    let file_flags = unsafe { FileFlowToFlags(flow, STREAM_TOSERVER) };
+    rs_nfs_setfileflags(0, state, file_flags);
+
+    if input.is_null() == true && input_len > 0 {
+        return rs_nfs_parse_request_tcp_gap(state, input_len);
+    }
```
Comment (inashivb) [inline rust/src/nfs/nfs.rs:1410]: Move all this except casts to separate commit

File: rust/src/nfs/nfs.rs [hunk 5]
```diff
@@ -1412,14 +1425,23 @@ pub extern "C" fn rs_nfs_parse_request_tcp_gap(
...
+                                        ) -> AppLayerResult
 {
+    let state = cast_pointer!(state, NFSState);
+    let flow = cast_pointer!(flow, Flow);
+    let file_flags = unsafe { FileFlowToFlags(flow, STREAM_TOCLIENT) };
+    rs_nfs_setfileflags(1, state, file_flags);
+
+    if input.is_null() == true && input_len > 0 {
+        return rs_nfs_parse_response_tcp_gap(state, input_len);
+    }
```
Comment (inashivb) [inline rust/src/nfs/nfs.rs:1444]: Move all this except casts to separate commit

File: rust/src/nfs/nfs.rs [hunk 6]
```diff
@@ -1438,46 +1460,56 @@ pub extern "C" fn rs_nfs_parse_response_tcp_gap(
...
                                        input_len: u32,
-                                       _data: *mut std::os::raw::c_void)
-                                       -> AppLayerResult
+                                       _data: *const std::os::raw::c_void,
+                                       _flags: u8) -> AppLayerResult
 {
+    let state = cast_pointer!(state, NFSState);
+    let file_flags = unsafe { FileFlowToFlags(f, STREAM_TOSERVER) };
+    rs_nfs_setfileflags(0, state, file_flags);
+
```
Comment (inashivb) [inline rust/src/nfs/nfs.rs:1474]: Move all this except casts to separate commit

File: rust/src/nfs/nfs.rs [hunk 7]
```diff
@@ -1438,46 +1460,56 @@ pub extern "C" fn rs_nfs_parse_response_tcp_gap(
...
                                         input_len: u32,
-                                        _data: *mut std::os::raw::c_void)
-                                        -> AppLayerResult
+                                        _data: *const std::os::raw::c_void,
+                                        _flags: u8) -> AppLayerResult
 {
+    let state = cast_pointer!(state, NFSState);
+    let file_flags = unsafe { FileFlowToFlags(f, STREAM_TOSERVER) };
+    rs_nfs_setfileflags(0, state, file_flags);
+
```
Comment (inashivb) [inline rust/src/nfs/nfs.rs:1492]: Move all this except casts to separate commit

File: rust/src/nfs/nfs.rs [hunk 8]
```diff
@@ -1566,6 +1607,18 @@ pub extern "C" fn rs_nfs_state_get_tx_detect_state(
...
+                                              event_type: *mut AppLayerEventType)
+                                             -> i8
+{
+    unsafe {
+        *event_name = b"NFS UDP event name (generic)\0".as_ptr() as *const std::os::raw::c_char;
+        *event_type = APP_LAYER_EVENT_TYPE_TRANSACTION;
+    }
+    0
+}
+
```
Comment (inashivb) [inline rust/src/nfs/nfs.rs:1621]: Move new fn addition to separate commit

File: rust/src/nfs/nfs.rs [hunk 9]
```diff
@@ -1799,57 +1855,93 @@ pub extern "C" fn rs_nfs_probe_ms(
...
+    if len < NFS_MIN_FRAME_LEN as u32 {
+        return ALPROTO_UNKNOWN;
+    }
     let slice: &[u8] = build_slice!(input, len as usize);
-    return nfs_probe_udp(slice, STREAM_TOSERVER);
+    match nfs_probe_udp(slice, STREAM_TOSERVER) {
+        1 => { return unsafe { ALPROTO_NFS } },
+        -1 => { return unsafe { ALPROTO_FAILED } },
+        _ => { return ALPROTO_UNKNOWN },
+    }
```
Comment (inashivb) [inline rust/src/nfs/nfs.rs:1917]: Remove return stmts

File: rust/src/nfs/nfs.rs [hunk 10]
```diff
@@ -1799,57 +1855,93 @@ pub extern "C" fn rs_nfs_probe_ms(
...
 {
+    if len < NFS_MIN_FRAME_LEN as u32 {
+        return ALPROTO_UNKNOWN;
+    }
     let slice: &[u8] = build_slice!(input, len as usize);
-    return nfs_probe_udp(slice, STREAM_TOCLIENT);
+    match nfs_probe_udp(slice, STREAM_TOCLIENT) {
+        1 => { return unsafe { ALPROTO_NFS } },
+        -1 => { return unsafe { ALPROTO_FAILED } },
+        _ => { return ALPROTO_UNKNOWN },
```
Comment (inashivb) [inline rust/src/nfs/nfs.rs:1936]: Remove return stmts

File: rust/src/nfs/nfs.rs (hunk 3 above)
Comment (inashivb) [reply rust/src/nfs/nfs.rs:1366]: @victorjulien is this check correct? Or should it be `if direction  == STREAM_TOCLIENT`? JFYI, `STREAM_TOCLIENT` is 0x08.

File: rust/src/nfs/nfs.rs (hunk 8 above)
Comment (victorjulien) [reply rust/src/nfs/nfs.rs:1621]: I actually think the whole event logic is unused in nfs, so lets not move it over

**Comment by victorjulien** [comment]
Seeing many failures in private tests.

**Comment by inashivb** [comment]
> Seeing many failures in private tests.

Have found some incorrect things. Will send a new PR.

**Comment by inashivb** [comment]
Replaced w #6232
